packages/tpetra/core/src/Tpetra_Details_residual.hpp

```diff
@@ -338,16 +338,18 @@ void localResidual(const CrsMatrix<SC,LO,GO,NO> &  A,
        (X_colmap_lcl.data () == B_lcl.data () && X_colmap_lcl.data () != nullptr),
        std::runtime_error, "X, Y and R may not alias one another.");
   }
-      
-#ifdef TPETRA_DETAILS_USE_REFERENCE_RESIDUAL
-  SC one = Teuchos::ScalarTraits<SC>::one();
-  SC negone = -one;
-  SC zero = Teuchos::ScalarTraits<SC>::zero();
-  // This is currently a "reference implementation" waiting until Kokkos Kernels provides
-  // a residual kernel.
-  A.localApply(X_colmap,R,Teuchos::NO_TRANS, one, zero);
-  R.update(one,B,negone);
-#else
+
+  const bool useNativeSpMVInResidual = ::Tpetra::Details::Behavior::useNativeSpMVInResidual ();
+  if (useNativeSpMVInResidual) {
+    SC one = Teuchos::ScalarTraits<SC>::one();
+    SC negone = -one;
+    SC zero = Teuchos::ScalarTraits<SC>::zero();
+    // This is currently a "reference implementation" waiting until Kokkos Kernels provides
+    // a residual kernel.
+    A.localApply(X_colmap,R,Teuchos::NO_TRANS, one, zero);
```

Comment:
I know this codepath predates your changes, but ``A.localApply`` is still going to use the default spmv (a TPL if available). If there's a need to use native spmv in Tpetra, it should probably be an optional parameter to localApply. When the KokkosKernels 4.3 promotion is merged, this path could pass in ``KokkosSparse::SPMV_NATIVE``). Then localApply would create a temporary handle if the algorithm is anything other than ``SPMV_DEFAULT``.

BTW, is a TPL implementation causing problems or showing bad performance? Or is this just about replacing a fixed macro with a Behavior option?

Reply:
I misunderstood the intent of this behavior, it really means "use a non-fused residual" so the codepath is correct for that.